File: js/dijkstra.js
export function findShortestPath(graph, source, target) {
	const done = {}     // {a: 0, b: 8: c: 9}
	const pending = {}  // {d: 17, e: 15}
	const prev = {}     // {a: null, b: "a", c: "a", d: "b", e: "c"}
	prev[source] = null
	done[source] = 0
	let current = source
	while(current !== null && current !== target) {
		const neighbours = graph.getNeighbours(+current)
		for(let {id, weight} of neighbours) {
			if(id in done) continue
			const distViaCurrent = done[current] + weight
			const isNew = !(id in pending)
			const isShorter = !isNew && distViaCurrent < pending[id]
			if(isNew || isShorter) {
				pending[id] = distViaCurrent
				prev[id] = current
			}
		}
		let next = findMinimum(pending)
		if(next) {
			done[next] = pending[next]
			delete pending[next]
		}
		current = next
	}
	if(target in done) {
		const route = joinRoute(target, prev)
		return {distance: done[target], route}
	}
	return "not reachable"
}

function findMinimum(pending) {
	let next = null
	let minDist = Infinity
	for(let id in pending) {
		if(pending[id] < minDist) {
			minDist = pending[id]
			next = id
		}
	}
	return next
}

function joinRoute(target, prev) {
	let previous = prev[target]
	let route = target
	while(previous !== null) {
		route = previous + "->" + route
		previous = prev[previous]
	}
	return route
}

File: python/dijkstra.py
def findShortestPath(graph, source, target):
	done = {}     # {"a": 0, "b": 8: "c": 9}
	pending = {}  # {"d": 17, "e": 15}
	prev = {}     # {"a": None, "b": "a", "c": "a", "d": "b", "e": "c"}
	prev[source] = None
	done[source] = 0
	current = source
	while(not current == None):
		neighbours = graph.getNeighbours(current)
		for neighbour in neighbours:
			nodeid = neighbour["id"]
			weight = neighbour["weight"]
			if nodeid in done:
				continue
			distViaCurrent = done[current] + weight
			isNew = not (nodeid in pending)
			isShorter = not(isNew) and distViaCurrent < pending[nodeid]
			if isNew or isShorter:
				pending[nodeid] = distViaCurrent
				prev[nodeid] = current
		nextnode = findMinimum(pending)
		if nextnode:
			done[nextnode] = pending[nextnode]
			del pending[nextnode]
		current = nextnode
	if target in done:
		route = joinRoute(target, prev)
		return {"distance": done[target], "route": route}
	return "not reachable"

def findMinimum(pending):
	nextnode = None
	minDist = float("inf")
	for id in pending:
		if(pending[id] < minDist):
			minDist = pending[id]
			nextnode = id
	return nextnode


def joinRoute(target, prev):
	previous = prev[target]
	route = str(target)
	while(previous != None):
		route = str(previous) + "->" + route
		previous = prev[previous]
	return route
